Create v1 of order flow documentation (#3)
Onyo/onyo.github.io: docs/checkout/orders/manage_orders.md

---
id: manage_orders
title: Fluxo de gerenciamento de pedidos
hide_title: true
---

## Autenticação para transações de pedidos

### Como funciona?

> Todas as requisições abaixo devem ser autenticadas. A autenticação deve ser fornecida através do envio do token da brand/company através do header HTTP Authorization.

Reviewer: ```suggestion
Todas as requisições abaixo requerem autenticação. A autenticação deve ser fornecida através do envio do token da brand/company através do header HTTP Authorization.
```

O token da brand/company pode ser fornecido diretamente pela Pede Pronto, usando o modelo JWT de autenticação.

**Resumo**

> O Pede Pronto disponibiliza o e-mail na listagem de pedidos, o PDV interpreta os dados, se ok prosegue alterando o status para o ID 4 (pos-accepted) ou ID 5 (pos-denied).

Reviewer: ```suggestion
O Pede Pronto disponibiliza o e-mail na listagem de pedidos, o PDV interpreta os dados, e se tudo estiver correto prosegue alterando o status para o ID 4 (pos-accepted) ou ID 5 (pos-denied).
```

### Listagem de pedidos (order-list)

> GET /v1/{:PARTNER_SLUG}/order

```
curl --request GET \
  --url https://pos.staging.onyo.com/v1/{:PARTNER_SLUG}/orders \
  --header 'Authorization: Bearer
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6Ikpva
G4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_ad
Qssw5c' \
  --header 'Content-Type: application/json'
```

> Essa chamada retorna uma lista de pedidos da loja autenticada na requisição. Por padrão essa chamada executa um filtro pelos pedidos nos status 2 e 3 (payment-authorized e pos-received) que são os pedidos que o PDV precisa aprovar ou reprovar. Caso algum filtro de status seja feito o filtro padrão será sobrescrito.

Reviewer: Talvez fosse interessante centralizar os status de pedidos comuns ao pos (só, pra deixar mais organizado).

Author: Tem uma tabela abaixo com todos os status. Seria isso?

| 3 | curbside | Um pedido a ser retirado no Drive Thru |
| 4 | table | Um pedido para ser servido na mesa informada |
| 5 | delivery | Um pedido a ser entregue no endereço informado |
| 6 | card | Um pedido para ser registrado em um cartão de consumo |
| 7 | payment | Um pagamento por itens encomendados pessoalmente |
| 8 | coupon | Um pedido que será resgatado na loja, mas que não deve ser preparado até que o cliente a resgate pessoalmente |
| 9 | room-service | Um pedido a ser entregue no local informado considerando algumas opções restritivas, como por exemplo um prédio comercial |

#### Campos Extras:

Os campos extras são retornados nessa chamada caso o pedido tenha informações adicionais como taxa de entrega ou descontos.

| **Campo** | **Descrição** |
| --------- | ------------- |
| optional | Campo booleano (true ou false) |
| value | Valor em R$ do extra/desconto |
| key | Campo descritivo do valor extra/desconto |
| label | Nome do valor extra/desconto |

Exemplo de resposta contendo um pedido:

```
{
 "pagination": {
   "next": null,
   "total": 1,
   "page": 1,
   "previous": null
 },

>  "data": [

Reviewer: Se der, da uma resumida nesse payload (só porque ele está muito grande). Se não, segue lá pelota

Author: Seria melhor, mas é bom deixar 'completo' do que depois virem perguntar o que aparece depois.

         "acquirer": {
           "pos_reference": "",
           "id": 1,
           "name": "Cielo"
         },
         "acquirer_return": {
           "Tid": "91744583",
           "Message": null,
           "ReturnCode": null,
           "authId": "744583"
         },
         "payment_method_pos_reference": "",
         "transaction_code": "TESTPAY1054575",
         "card": {
           "masked_number": "4444331111",
           "printed_name": "Felipe",
           "card_brand": {
             "pos_reference": "",
             "code": "visa",
             "id": 8,
             "name": "Visa"
           }
         }
       }
     ],
     "error": null,
     "discount_value": "7.35",
     "sequence_number": null
   }
 ]
}

```

### Detalhe de pedido (order-detail)

> GET /v1/{:PARTNER_SLUG}/order/{:order_id}

```
curl --request GET \
  --url https://pos.staging.onyo.com/v1/{:PARTNER_SLUG}/orders/1234 \
  --header 'Authorization: Bearer
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6Ikpva
G4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_ad
Qssw5c' \
  --header 'Content-Type: application/json'
```

> Essa chamada retorna o JSON do pedido informado na URL. A resposta dessa chamada é o mesmo conteúdo da lista de itens da chamada de listagem de pedidos, porém apenas com o objeto do pedido solicitado na URL. 

Reviewer: ```suggestion
Essa chamada retorna o JSON do pedido informado na URL. A resposta dessa chamada tem o mesmo conteúdo da lista de itens da chamada de listagem de pedidos, porém apenas com o objeto do pedido solicitado na URL. 
```

### Atualização de pedido (order-patch)

> PATCH /v1/{:PARTNER_SLUG}/order/{:order_id}

Essa chamada permite a atualização do status de um pedido para 4 (aceito / POS_ACCEPTED) ou 5 (erro / POS_DENIED).
Caso haja algum erro com o pedido é importante que o campo erro seja enviado junto com o status 5.

Exemplo de payload de aprovação de pedido:

```
{
    "status": 4
}
```

** Status disponíveis para alteração **

Um status não pode retroceder, a não ser para 0 (Cancelled) ou 5 (POS_DENIED).

**Exemplo correto:** ```1 (backend-received) > 2 (payment-authorized) > 3 (pos-received) > 4 (pos-accepted) > 6 (read) > 7 (dispatched) ```

**Exemplo inválido:** ``` 1 (backend-received) > 2 (payment-authorized) > 3 (pos-received) > 4 (pos-accepted) > 3 (pos-received) ```

> | **ID** | **Código (Inteiro)** | **Descrição** |

Reviewer: Acho que essa tabela poderia ficar lá em cima. O que acha?

| ------ | -------------------- | ------------- |
| 0 |canceled |Cancelado devido ao pagamento ter sido negado ou cancelado manualmente pela Pede Pronto |
| 1 | backend-received | Aguardando processamento do pagamento |
| 2 | payment-authorized | Pagamento aprovado, aguardando processamento pelo PDV |
| 3 | pos-received | Recebido pelo PDV, aguardando confirmação |
| 4 | pos-accepted | Aceito pelo PDV |
| 5 | pos-denied | Negado pelo PDV |
| 6 | ready | Pedido pronto para ser recebido pelo consumidor (informado pela loja) |
| 7 | dispatched | Saiu para entrega (delivery) |
| 8 | delivered | Entregue para o consumidor|
| 9 | received | Recebido pelo consumidor (informado pelo próprio consumidor) |
| 13 | preparing | Pedido sendo preparado pela loja |
| 15 | product-unavailable | Preparo do pedido pausado devido a indisponibilidade de algum produto. Aguardando decisão do consumidor |
| 16 | customer-action-needed | Preparo do pedido pausado devido a uma dúvida do Staff da loja. Aguardando informação do consumidor |

> **Status não aplicáveis para integrações deste manual**

Reviewer: Se não é aplicável, talvez valha a pena retirar.

Author: Pois é, pensei mesma coisa, porém para deixar 'tudo documentado' preferi manter isso. O que acha?

Reviewer: Bele, segue o baile
